feat(open-api): add task endpoints to the open api

koevskinikola opened the pull request:
[![CAM-11548](https://badgen.net/badge/JIRA/CAM-11548/0052CC)](https://app.camunda.com/jira/browse/CAM-11548)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

* Add the Task endpoints (without sub-resources) to the Open API spec. They include: querying for tasks, counting, create, update, get,
complete, assign, resolve, claim, unclaim, delegate, get form object,
deployed/rendered form, submit form, get form variables, throw bpmn
errors and escalations.

Related to CAM-11548

koevskinikola commented on engine-rest/engine-rest-openapi/src/main/templates/models/org/camunda/bpm/engine/rest/dto/runtime/task/TaskQueryDto.ftl:
```diff
@@ -0,0 +1,655 @@
+        name = "dueDate"
+        type = "string"
+        format = "date-time"
+        desc = "Restrict to tasks that are due on the given date. By default*, the date must have the
```
@yanavasileva, I forgot to add links to the default format in this file. Please reopen the ticket once you review it, so I can add `[default](${docsUrl}/reference/rest/overview/date-format/)`.

yanavasileva commented on engine-rest/engine-rest-openapi/src/main/templates/models/org/camunda/bpm/engine/rest/dto/runtime/task/TaskQueryDto.ftl:
```diff
@@ -0,0 +1,655 @@
+                A process instance matches a nested query if it fulfills at least one of the query's predicates.
+                With multiple nested queries, a process instance must fulfill at least one predicate of each query (Conjunctive Normal Form).
+                All process instance query properties can be used except for: `sorting`
+                See the [User guide](${docsUrl}/user-guide/process-engine/process-engine-api/#or-queries) for more information about OR queries." />
```
dto and description are wrong, please double check the line endings and formatting
```suggestion
        dto = "TaskQueryDto"
        desc = "A JSON array of nested task queries with OR semantics. A task matches a nested query if it fulfills
                *at least one* of the query's predicates. With multiple nested queries, a task must fulfill at least one predicate of *each* query ([Conjunctive Normal Form](https://en.wikipedia.org/wiki/Conjunctive_normal_form)).

                All task query properties can be used except for: `sorting`, `withCandidateGroups`,
    `withoutCandidateGroups`, `withCandidateUsers`, `withoutCandidateUsers`

                See the [User guide](${docsUrl}/user-guide/process-engine/process-engine-api/#or-queries)
    for more information about OR queries." />
```

yanavasileva commented on engine-rest/engine-rest-openapi/src/main/templates/models/org/camunda/bpm/engine/rest/dto/runtime/task/TaskQueryDto.ftl:
```diff
@@ -0,0 +1,655 @@
+      "items":
+
+        <#assign last = true >
+        <#assign sortByValues = ['"instanceId"', '"definitionId"', '"definitionKey"', '"businessKey"', '"tenantId"']>
```
`sortByValues` are wrong:
instanceId, caseInstanceId, dueDate, executionId, caseExecutionId,assignee, created, description, id, name, nameCaseInsensitive, priority, processVariable, executionVariable, taskVariable, caseExecutionVariable, and caseInstanceVariable

yanavasileva commented on engine-rest/engine-rest-openapi/src/main/templates/models/org/camunda/bpm/engine/rest/dto/runtime/task/UserIdDto.ftl:
```diff
@@ -0,0 +1,12 @@
+        name = "userId"
+        type = "string"
+        last =  true
+        desc = "The id of the user that ${action}." />
```
Please generalize this and add descriptions to the `requestBody`-ies where this DTO is used.

yanavasileva commented on engine-rest/engine-rest-openapi/src/main/templates/paths/task/{id}/complete/post.ftl:
```diff
@@ -0,0 +1,69 @@
+  "responses" : {
+
+    <@lib.response
+        code = "200"
```
please specify DTO

yanavasileva commented on engine-rest/engine-rest-openapi/src/main/templates/paths/task/{id}/complete/post.ftl:
```diff
@@ -0,0 +1,69 @@
+
+    <@lib.response
+        code = "200"
+        desc = "Request successful. The response contains the process variables." />
```
please add an example

yanavasileva commented on engine-rest/engine-rest-openapi/src/main/templates/paths/task/{id}/complete/post.ftl:
```diff
@@ -0,0 +1,69 @@
+        desc = "Request successful. The response contains the process variables." />
+
+    <@lib.response
+        code = "204"
```
please add an example

koevskinikola replied:
This response doesn't return any content. I will add an example for it in the HTTP Request.

yanavasileva replied:
👍

yanavasileva commented on engine-rest/engine-rest-openapi/src/main/templates/paths/task/{id}/deployed-form/get.ftl:
```diff
@@ -0,0 +1,76 @@
+
+    <@lib.response
+        code = "200"
+        mediaType = "application/xhtml+xml"
```
the `ExceptionDto` is still referenced in response by default, we should fix the macro

yanavasileva commented on engine-rest/engine-rest-openapi/src/main/templates/paths/task/{id}/rendered-form/get.ftl:
```diff
@@ -0,0 +1,57 @@
+
+    <@lib.response
+        code = "200"
+        mediaType = "application/xhtml+xml"
```
the ExceptionDto is still referenced in response by default, we should fix the macro

yanavasileva commented on engine-rest/engine-rest-openapi/src/main/templates/paths/task/{id}/submit-form/post.ftl:
```diff
@@ -0,0 +1,126 @@
+  "responses" : {
+
+    <@lib.response
+        code = "200"
```
please  specify dto

koevskinikola commented:
@yanavasileva I also think we should prefix all Task-related endpoint `tags` with `Task`. E.g. `Task Attachment`, `Task Comment`... So that they are grouped in the UI in one continuous section. Let's discuss this.

koevskinikola commented:
To respond to the first comment, yes

>     * delete task is missing in docs, could you please check if we need to add it to docs(manual and openapi)?

Yes, we definitely need to add it to both places. For the docs, it maps to this ticket: https://jira.camunda.com/browse/CAM-8239

koevskinikola commented:
@yanavasileva, IDid the following:
1.  Implemented the review hints;
2. Added the missing `create` and `delete` endpoints
3. Adjusted the `response` macro to handle `application/xhtml+xml` responses. Unfortunately, we currently can't properly format Response types with multiple `mediaTypes` since it's hard to distinguish which example belongs to which `mediaType`. We will have to add them manually, but it's a low occurrence, I think.
